### ISSUE TITLE
$PATH for interactive non-login shells is sadly limited

### ISSUE THREAD

**Issue by miabbott** [openshift/os#191]
While debugging `coreos.selinux.enforcing` on RHCOS, the test was reporting `command not found` for `getenforce`.  Yet when I logged in directly or via `Manhole()`, I was able to run `getenforce` just fine.

When I finally inspected the value of `$PATH` in both cases, I found the issue.  When logged in directly (as `core` user), I get the following `$PATH`

`PATH=/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/core/.local/bin:/home/core/bin`

...but when running a test via `kola`:

`PATH=/usr/local/bin:/usr/bin`

And since `getenforce` lives in `/usr/sbin`, the `kola` tests are unable to find it during execution.

I think I was able to make sense of how `$PATH` was getting set for the login shell case (thanks `man bash`!), but I haven't been able to figure out where the default path is configured for the non-login case.

Related to this, the value for `$PATH` in the non-login case on Container Linux looks like this:

`PATH=/usr/bin:/bin:/usr/sbin:/sbin`

**Comment by ashcrow** [comment]
This may be an artifact on how `kola` is testing. I'll take a look from that end.

**Comment by ashcrow** [comment]
@arithx noted that `kolet` isn't used in this case.

**Comment by ashcrow** [comment]
@miabbott noted that by adding in  `/bin/bash -l -c getenforce` as part of commands the `PATH` was set properly. He's going to give `session.SetEnv` a go and see if that is a bit cleaner.

**Comment by miabbott** [comment]
Strangely, if I try setting the `PATH` using `session.Setenv()` for all the SSH connections, I get a failure trying to `setenv`.

Here's the diff I used:

```diff
$ git diff platform/base.go
diff --git a/platform/base.go b/platform/base.go
index 77f1fa0d..0a2e11fa 100644
--- a/platform/base.go
+++ b/platform/base.go
@@ -119,6 +119,10 @@ func (bc *BaseCluster) SSH(m Machine, cmd string) ([]byte, []byte, error) {
        }
        defer session.Close()
 
+       // explicitly set PATH because RHCOS has weird defaults
+       if err := session.Setenv("PATH", "/usr/bin:/bin:/usr/sbin:/sbin"); err != nil {
+               return nil, nil, err
+       }
        session.Stdout = &stdout
        session.Stderr = &stderr
        err = session.Run(cmd)
```

And the failure from the harness:

```
$ bin/kola run -d -p aws $(rhcos) coreos.selinux.enforce                                                                                                 
2018-07-27T18:30:21Z cli: Started logging at level DEBUG
=== RUN   coreos.selinux.enforce
--- FAIL: coreos.selinux.enforce (401.18s)
        harness.go:429: Cluster failed starting machines: machine "i-0910ca1661c72c31b" failed basic checks: ssh unreachable: could not check if machine is running: : ssh: setenv failed:                        
FAIL, output in _kola_temp/aws-2018-07-27-1830-3742
harness: test suite failed
```

Note, this isn't blocking any progress per se, but it feels like we should be providing more sane defaults on non-login shells.

**Comment by bgilbert** [comment]
Is `PATH` in sshd's `AcceptEnv` setting?

**Comment by miabbott** [comment]
> Is PATH in sshd's AcceptEnv setting?

Nope.

```bash
$ sudo grep AcceptEnv /etc/ssh/sshd_config 
AcceptEnv LANG LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES
AcceptEnv LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT
AcceptEnv LC_IDENTIFICATION LC_ALL LANGUAGE
AcceptEnv XMODIFIERS
```

**Comment by miabbott** [comment]
I tried the `AcceptEnv` route but just hit more walls.

I appended `AcceptEnv PATH` to `/etc/ssh/sshd_config` and then tried to pass the value along via the `SendEnv` option to `ssh`, but still got the same value as before:

```bash
$ env PATH=/usr/bin:/bin:/usr/sbin:/sbin ssh -l core -o SendEnv=PATH 10.8.245.216 printenv
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
XDG_SESSION_ID=16
SELINUX_ROLE_REQUESTED=
SHELL=/bin/bash
SSH_CLIENT=10.3.117.59 50580 22
SELINUX_USE_CURRENT_RANGE=
USER=core
MAIL=/var/mail/core
PATH=/usr/local/bin:/usr/bin
PWD=/var/home/core
XMODIFIERS=@im=ibus
LANG=en_US.UTF-8
SELINUX_LEVEL_REQUESTED=
SHLVL=1
HOME=/home/core
LOGNAME=core
SSH_CONNECTION=10.3.117.59 50580 172.16.171.3 22
LESSOPEN=||/usr/bin/lesspipe.sh %s
XDG_RUNTIME_DIR=/run/user/1000
_=/usr/bin/printenv
```

Even tried disabling SELinux completely as a hail mary, but nothing changed.

Will continue to explore other options..

**Comment by miabbott** [comment]
Here's a potential solution...

We drop a file in `/etc/profile.d` that sets up the `PATH` variable

```bash
# cat /etc/profile.d/path.sh 
PATH=/usr/bin:/bin:/usr/sbin:/sbin
```

This seems to be a relatively safe change; it's solving our problem and doesn't appear to break existing behaviors.

**Without `/etc/profile.d/path.sh`**

```
$ ssh -l root 10.8.245.216
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
Last login: Tue Jul 31 17:04:14 2018 from ovpn-117-59.phx2.redhat.com
Red Hat CoreOS
  Information: https://url.corp.redhat.com/redhat-coreos
  Bugs: https://github.com/openshift/os
---

[root@host-172-16-171-3 ~]# echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/.local/bin:/root/bin
[root@host-172-16-171-3 ~]# exit
logout
Connection to 10.8.245.216 closed.

$ ssh -l root 10.8.245.216 printenv | grep PATH
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin

$ ssh -l core 10.8.245.216
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
Last login: Tue Jul 31 17:01:23 2018 from ovpn-117-59.phx2.redhat.com
Red Hat CoreOS
  Information: https://url.corp.redhat.com/redhat-coreos
  Bugs: https://github.com/openshift/os
---

[core@host-172-16-171-3 ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/core/.local/bin:/home/core/bin
[core@host-172-16-171-3 ~]$ exit
logout
Connection to 10.8.245.216 closed.

$ ssh -l core 10.8.245.216 printenv | grep PATH
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
PATH=/usr/local/bin:/usr/bin
```

**With `/etc/profile.d/path.sh`**

```
$ ssh -l root 10.8.245.216
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
Last login: Tue Jul 31 17:08:22 2018 from ovpn-117-59.phx2.redhat.com
Red Hat CoreOS
  Information: https://url.corp.redhat.com/redhat-coreos
  Bugs: https://github.com/openshift/os
---

[root@host-172-16-171-3 ~]# echo $PATH
/usr/bin:/bin:/usr/sbin:/sbin:/root/.local/bin:/root/bin
[root@host-172-16-171-3 ~]# exit
logout
Connection to 10.8.245.216 closed.

$ ssh -l root 10.8.245.216 printenv | grep PATH
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
PATH=/usr/bin:/bin:/usr/sbin:/sbin

$ ssh -l core 10.8.245.216
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
Last login: Tue Jul 31 17:07:19 2018 from ovpn-117-59.phx2.redhat.com
Red Hat CoreOS
  Information: https://url.corp.redhat.com/redhat-coreos
  Bugs: https://github.com/openshift/os
---

[core@host-172-16-171-3 ~]$ echo $PATH
/usr/bin:/bin:/usr/sbin:/sbin:/home/core/.local/bin:/home/core/bin
[core@host-172-16-171-3 ~]$ exit
logout
Connection to 10.8.245.216 closed.

$ ssh -l core 10.8.245.216 printenv | grep PATH
Warning: Permanently added '10.8.245.216' (ECDSA) to the list of known hosts.
PATH=/usr/bin:/bin:/usr/sbin:/sbin
```

In the example above, I used the `PATH` value used by CL.  We could use whatever value we want for RHCOS.

@ashcrow  @cgwalters WDYT?

**Comment by ashcrow** [comment]
@miabbott :+1: I'm happy with the file in ``/etc/profile.d``.

**Comment by cgwalters** [comment]
I vote we just unify `/bin` and `/sbin` on RHCOS - and push for doing so in Fedora CoreOS too.

**Comment by cgwalters** [comment]
Last I checked the only thing stopping that on Fedora+derivatives was the Fedora-specific [usermode](https://src.fedoraproject.org/rpms/usermode) program, and that's only used by things we generally don't care about on the host like `mock`.

**Comment by cgwalters** [comment]
FWIW dropping the profile entry seems OK to me - no reason not to do that, though I have a WIP patch ready for sbin unification I'm testing now.

**Comment by miabbott** [comment]
This was fixed with openshift/redhat-release-coreos#8
